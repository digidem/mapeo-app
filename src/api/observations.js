// @flow
import fetch from 'fetch';
import { jsonRequest } from './base';

export interface CreateRequest {
  lat: number;
  lon: number;
}

export type UpdateRequest = {
  id: string
};

class Observation {
  static list = () =>
    jsonRequest({
      method: 'GET',
      route: '/observations'
    });

  static create = (observation: CreateRequest) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: {
        device_id: '1',
        ...observation
      }
    });

  static update = (observation: UpdateRequest) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: {
        ...observation
      }
    });
}

export default Observation;
