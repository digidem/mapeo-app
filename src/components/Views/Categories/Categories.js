// @flow
import React from 'react';
import {
  Dimensions,
  FlatList,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native';
import { NavigationActions } from 'react-navigation';
import I18n from 'react-native-i18n';
import CloseIcon from 'react-native-vector-icons/MaterialCommunityIcons';
import CheckIcon from 'react-native-vector-icons/Octicons';
import Image from 'react-native-remote-svg';

import type { Category } from '../../../types/category';
import type { Field } from '../../../types/field';
import type { Observation } from '../../../types/observation';
import Header from '../../Base/Header';
import {
  DARK_GREY,
  LIGHT_GREY,
  VERY_LIGHT_BLUE,
  WHITE
} from '../../../lib/styles';

type Props = {
  navigation: NavigationActions
};

export type StateProps = {
  allFields: Object,
  categories: Category[],
  selectedObservation?: Observation,
  updateFlow: boolean,
  icons: Object
};

export type DispatchProps = {
  listCategories: () => void,
  updateObservation: (obs: Object) => void,
  clearSelectedObservation: () => void
};

const styles = StyleSheet.create({
  header: {
    flexDirection: 'row',
    flex: 1,
    backgroundColor: DARK_GREY,
    paddingVertical: 20,
    alignItems: 'center',
    justifyContent: 'center'
  },
  close: { position: 'absolute', left: 10 },
  title: { fontSize: 20, fontWeight: '700', color: LIGHT_GREY },
  circle: {
    width: 50,
    height: 50,
    backgroundColor: 'white',
    borderRadius: 50,
    borderColor: LIGHT_GREY,
    borderWidth: 1,
    justifyContent: 'center',
    alignItems: 'center',
    shadowColor: 'black',
    shadowRadius: 10,
    shadowOpacity: 1,
    shadowOffset: { width: 0, height: 2 },
    elevation: 3,
    marginBottom: 5
  },
  innerCircle: {
    width: 40,
    height: 40,
    backgroundColor: LIGHT_GREY,
    borderRadius: 50,
    alignItems: 'center',
    justifyContent: 'center'
  },
  cellContainer: {
    flex: 1,
    height: 130,
    paddingTop: 20,
    paddingBottom: 20,
    width: Dimensions.get('window').width / 3
  },
  cell: { flex: 1, flexDirection: 'column', alignItems: 'center' },
  categoryName: {
    color: 'black',
    fontWeight: '700',
    textAlign: 'center',
    maxWidth: Dimensions.get('window').width / 3 - 2
  },
  checkIcon: { alignSelf: 'center', marginLeft: 3 },
  greyCheck: {
    backgroundColor: LIGHT_GREY,
    height: 25,
    width: 25,
    borderRadius: 50,
    justifyContent: 'center'
  },
  greyCheckOuterCircle: {
    width: 30,
    height: 30,
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#d6d2cf'
  }
});

I18n.fallbacks = true;
I18n.translations = {
  en: require('../../../translations/en'),
  es: require('../../../translations/es')
};

class Categories extends React.Component<Props & StateProps & DispatchProps> {
  componentDidMount() {
    const { listCategories } = this.props;

    listCategories();
  }

  shouldComponentUpdate(nextProps: Props & StateProps & DispatchProps) {
    if (nextProps.navigation.isFocused()) {
      return nextProps !== this.props;
    }

    return false;
  }

  map: any;
  handleUpdateObservation = (item: Category) => {
    const {
      allFields,
      updateObservation,
      selectedObservation,
      navigation
    } = this.props;

    if (selectedObservation) {
      updateObservation({
        id: selectedObservation.id,
        categoryId: item.id,
        fields: item && item.fields.map(field => allFields[field])
      });

      navigation.navigate({
        routeName: 'ObservationEditor',
        params: {
          category: item.id
        }
      });
    }
  };

  renderItem = ({ item }) => {
    const { icons } = this.props;

    return (
      <TouchableOpacity
        style={styles.cellContainer}
        underlayColor="transparent"
        onPress={() => {
          this.handleUpdateObservation(item);
        }}
      >
        <View style={styles.cell}>
          {!!item.icon &&
            !!icons[item.icon] && (
              <View style={styles.circle}>
                <Image
                  source={{
                    uri: `data:image/svg+xml;utf8,${icons[item.icon]}`
                  }}
                  height="30"
                  width="30"
                  style={{ height: 30, width: 30 }}
                />
              </View>
            )}
          <Text style={styles.categoryName}>{item.name}</Text>
        </View>
      </TouchableOpacity>
    );
  };

  handleBack = () => {
    const { navigation, clearSelectedObservation, updateFlow } = this.props;

    if (!updateFlow) {
      clearSelectedObservation();
    }

    navigation.goBack();
  };

  render() {
    const { categories } = this.props;
    const keyExtractor = item => item.id;

    return (
      <View style={{ flex: 1 }}>
        <FlatList
          ListHeaderComponent={
            <Header
              leftIcon={
                <TouchableOpacity
                  underlayColor="rgba(0, 0, 0, 0.5)"
                  onPress={this.handleBack}
                >
                  <CloseIcon color="#9E9C9C" name="window-close" size={30} />
                </TouchableOpacity>
              }
              rightIcon={
                <View style={styles.greyCheckOuterCircle}>
                  <View style={styles.greyCheck}>
                    <CheckIcon
                      color="white"
                      name="check"
                      size={18}
                      style={styles.checkIcon}
                    />
                  </View>
                </View>
              }
            />
          }
          style={{
            height: 80,
            width: Dimensions.get('window').width,
            backgroundColor: VERY_LIGHT_BLUE
          }}
          keyExtractor={keyExtractor}
          renderItem={this.renderItem}
          data={categories}
          numColumns={3}
        />
      </View>
    );
  }
}

export default Categories;
