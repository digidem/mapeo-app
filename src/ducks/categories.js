// @flow

import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { create } from '../lib/redux';

export const {
  type: CATEGORY_LIST,
  action: categoryList,
  reducer: categoryListReducer
} = create('CATEGORY_LIST', {
  success: (state, meta, payload) => {
    const newState = update(state, {
      categories: { $set: keyBy(payload, 'id') }
    });

    return newState;
  }
});

export default [categoryListReducer];
