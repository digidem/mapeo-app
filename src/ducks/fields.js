// @flow

import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { create } from '../lib/redux';

export const {
  type: FIELD_LIST,
  action: fieldList,
  reducer: fieldListReducer
} = create('FIELD_LIST', {
  success: (state, action) => {
    const newState = update(state, {
      fields: { $set: keyBy(action.payload, 'id') }
    });

    return newState;
  }
});

export const {
  type: PRESETS_SELECT,
  action: presetsSelect,
  reducer: presetsSelectReducer
} = create('PRESETS_SELECT', {});

export const {
  type: PRESETS_ICONS_LIST,
  action: presetsIconsList,
  reducer: presetsIconsListReducer
} = create('PRESETS_ICONS_LIST', {
  success: (state, action) =>
    update(state, {
      icons: { $set: action.payload }
    })
});

export default [
  fieldListReducer,
  presetsSelectReducer,
  presetsIconsListReducer
];
