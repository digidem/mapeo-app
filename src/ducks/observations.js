// @flow

import update from 'immutability-helper';
import {keyBy} from 'lodash';
import {create} from '../lib/redux';
import {createSelector} from '../lib/selector';
import type {StoreState} from '../types/redux';
import type {Observation} from '../types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const observations = keyBy(action.payload, 'id');

    if (state.observations) {
      Object.keys(state.observations).forEach(id => observations[id] = {
        ...observations[id],
        ...state.observations[id],
      });
    }

    const newState = update(state, {observations: {$set: observations}});

    return newState;
  }
});

export const {
  type: OBSERVATION_CREATE,
  action: observationCreate,
  reducer: observationCreateReducer
} = create('OBSERVATION_CREATE', {
  start: (state, action) => {
    const observation = action.meta;

    if (!observation.lat && !observation.lon && state.gps.data) {
      observation.lat = state.gps.data.latitude;
      observation.lon = state.gps.data.longitude;
    }

    const newState = update(state, {selectedObservation: {$set: observation}});

    return newState;
  },
  success: (state, action) => {
    const newState = update(state, {selectObservation: {$set: action.payload}});

    return newState;
  }
});

export const {
  type: OBSERVATION_UPDATE,
  action: observationUpdate,
  reducer: observationUpdateReducer
} = create('OBSERVATION_UPDATE', {
  start: (state, action) => {
    let newState;
    if (state.selectedObservation &&
        action.meta.id === state.selectedObservation.id) {
      newState = update(state, {
        selectedObservation: {
          $set: {...state.selectedObservation, ...action.meta},
        }
      });
    } else {
      newState = update(state, {
        observations: {
          [action.meta.id]: {
            $set: {...state.observations[action.meta.id], ...action.meta}
          }
        }
      });
    }

    return newState;
  }
});

export const {
  type: OBSERVATION_SELECT,
  action: observationSelect,
  reducer: observationSelectReducer
} = create('OBSERVATION_SELECT', {
  start: (state, action) => {
    if (action.meta === undefined && state.selectedObservation) {
      return update(state, {
        selectedObservation: {$set: action.meta},
        observations: {
          [state.selectedObservation.id]: {
            $set: {
              ...(state.observations[state.selectedObservation.id] || {}),
              ...state.selectedObservation,
            }
          }
        }
      });
    } else {
      return update(state, {selectedObservation: {$set: action.meta}});
    }
  }
});

export const selectObservation = createSelector(
    [(state: StoreState, id: string): Observation =>
         state.app.observations[id]],
    (observation: Observation): Observation => observation,
    (observation: Observation, id: string): string => id);

export default [
  observationListReducer, observationCreateReducer, observationUpdateReducer,
  observationSelectReducer
];
