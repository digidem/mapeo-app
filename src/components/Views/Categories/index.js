// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import { values } from 'lodash';
import { withNavigationFocus } from 'react-navigation';
import { StoreState } from '../../../types/redux';
import { categoryList } from '../../../ducks/categories';
import {
  observationUpdate,
  observationSelect
} from '../../../ducks/observations';
import { observationSource } from '../../../ducks/observationSource';
import Categories from './Categories';
import type { StateProps } from './Categories';

function mapStateToProps(state: StoreState): StateProps {
  const categories = values(state.app.categories).sort(
    (a, b) => a.name - b.name
  );
  const { selectedObservation, observations } = state.app;

  let updateFlow = false;
  if (selectedObservation && observations[selectedObservation.id]) {
    updateFlow = true;
  }

  return {
    allFields: state.app.fields,
    categories,
    selectedObservation,
    updateFlow,
    icons: state.app.icons
  };
}

function mapDispatchToProps(dispatch: Dispatch) {
  return {
    updateObservation: observation => dispatch(observationUpdate(observation)),
    listCategories: () => {
      dispatch(categoryList(''));
    },
    clearSelectedObservation: () => {
      dispatch(observationSelect(undefined));
      dispatch(observationSource(undefined));
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(
  withNavigationFocus(Categories)
);
