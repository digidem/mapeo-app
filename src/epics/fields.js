// @flow
import React from 'react';
import { Observable } from 'rxjs';
import type { ActionsObservable } from 'redux-observable';
import {
  PRESETS_SELECT,
  presetsSelect,
  fieldList,
  presetsIconsList,
  PRESETS_ICONS_LIST
} from '../ducks/fields';
import { categoryList } from '../ducks/categories';
import type { Action } from '../types/redux';
import type { Field } from '../types/field';
import { values } from 'lodash';
import Presets from '../api/presets';
import DOMParser from 'react-native-xml2js';

export const presetsSelectEpic = (
  action$: ActionsObservable<Action<string, Field[]>>
) =>
  action$
    .ofType(PRESETS_SELECT)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Presets.get(action.meta).flatMap(presets =>
        Observable.concat(
          Observable.of(
            fieldList(
              '',
              Object.keys(presets.fields).map(k => ({
                name: presets.fields[k].key,
                id: k,
                type: presets.fields[k].type,
                placeholder: presets.fields[k].placeholder,
                answered: false,
                answer: ''
              }))
            )
          ),
          Observable.of(
            categoryList(
              '',
              Object.keys(presets.presets).map(k => ({
                name: presets.presets[k].name,
                id: k,
                icon: presets.presets[k].icon,
                fields: presets.presets[k].fields
              }))
            )
          ),
          Observable.of(presetsIconsList(action.meta))
        )
      )
    );

export const presetsIconsListEpic = (
  action$: ActionsObservable<Action<string, Field[]>>
) =>
  action$
    .ofType(PRESETS_ICONS_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Presets.icons(action.meta).flatMap(iconsString =>
        Observable.create(observer => {
          DOMParser.parseString(iconsString, { trim: true }, (err, result) => {
            if (err) {
              observer.error(err);
            }

            const icons = {};
            const builder = new DOMParser.Builder({
              rootName: 'svg',
              headless: true
            });
            result.svg.symbol.forEach(s => {
              icons[s.$.id] = builder.buildObject(s);
            });

            observer.next(icons);
          });
        })
          .map(icons => presetsIconsList(action.meta, icons))
          .catch(err => Observable.of(presetsIconsList(action.meta, err)))
      )
    );

export default [presetsSelectEpic, presetsIconsListEpic];
