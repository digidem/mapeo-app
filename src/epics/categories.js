// @flow
import React from 'react';
import type { ActionsObservable } from 'redux-observable';
import { Image } from 'react-native';
import { CATEGORY_LIST, categoryList } from '../ducks/categories';
import { PRESETS_SELECT, presetsSelect } from '../ducks/fields';
import { createCategory } from '../mocks/categories';
import type { Action } from '../types/redux';
import type { Category } from '../types/category';
import Presets from '../api/presets';

export const categoryListEpic = (
  action$: ActionsObservable<Action<string, Category[]>>
) =>
  action$
    .ofType(CATEGORY_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() => Presets.list().map(list => presetsSelect(list[0])));

export default [categoryListEpic];
