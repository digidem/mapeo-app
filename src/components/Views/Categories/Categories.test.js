// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import Categories from './Categories';
import { createObservation } from '../../../mocks/observations';
import { createCategory } from '../../../mocks/categories';
import { createField } from '../../../mocks/fields';

describe('Categories tests', () => {
  const addListener = () => true;
  const categories = [createCategory()];
  const selectedObservation = createObservation();
  const listCategories = jest.fn();
  const updateObservation = jest.fn();
  const clearSelectedObservation = jest.fn();
  const field = createField();
  const icons = {
    [createCategory().icon]: 'icon'
  };

  test('snapshot', () => {
    const tree = renderer
      .create(
        <Categories
          categories={categories}
          navigation={{}}
          selectedObservation={selectedObservation}
          listCategories={listCategories}
          updateObservation={updateObservation}
          allFields={{ [field.id]: field }}
          updateFlow={false}
          clearSelectedObservation={clearSelectedObservation}
          icons={icons}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
