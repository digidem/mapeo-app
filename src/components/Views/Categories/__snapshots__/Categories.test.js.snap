// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories tests snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListHeaderComponent={
      <Unknown
        leftIcon={
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
            underlayColor="rgba(0, 0, 0, 0.5)"
          >
            <Icon
              allowFontScaling={false}
              color="#9E9C9C"
              name="window-close"
              size={30}
            />
          </TouchableOpacity>
        }
        rightIcon={
          <Component
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#d6d2cf",
                "borderRadius": 50,
                "height": 30,
                "justifyContent": "center",
                "width": 30,
              }
            }
          >
            <Component
              style={
                Object {
                  "backgroundColor": "#EAEAEA",
                  "borderRadius": 50,
                  "height": 25,
                  "justifyContent": "center",
                  "width": 25,
                }
              }
            >
              <Icon
                allowFontScaling={false}
                color="white"
                name="check"
                size={18}
                style={
                  Object {
                    "alignSelf": "center",
                    "marginLeft": 3,
                  }
                }
              />
            </Component>
          </Component>
        }
      />
    }
    data={
      Array [
        Object {
          "fields": Array [],
          "icon": Object {
            "testUri": "../../../src/images/categories/category_18.png",
          },
          "id": "categoryID",
          "name": "categoryName",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    invertStickyHeaders={undefined}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#ccffff",
        "height": 80,
        "width": 750,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              undefined,
            ]
          }
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "height": 130,
                "opacity": 1,
                "paddingBottom": 20,
                "paddingTop": 20,
                "width": 250,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "white",
                    "borderColor": "#EAEAEA",
                    "borderRadius": 50,
                    "borderWidth": 1,
                    "elevation": 3,
                    "height": 50,
                    "justifyContent": "center",
                    "marginBottom": 5,
                    "shadowColor": "black",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 1,
                    "shadowRadius": 10,
                    "width": 50,
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "height": 30,
                        "width": 30,
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <RCTWebView
                      allowsInlineMediaPlayback={undefined}
                      automaticallyAdjustContentInsets={undefined}
                      bounces={undefined}
                      contentInset={undefined}
                      dataDetectorTypes={undefined}
                      decelerationRate={undefined}
                      injectedJavaScript={undefined}
                      mediaPlaybackRequiresUserAction={undefined}
                      messagingEnabled={false}
                      onLoadingError={[Function]}
                      onLoadingFinish={[Function]}
                      onLoadingStart={[Function]}
                      onMessage={[Function]}
                      onShouldStartLoadWithRequest={undefined}
                      scalesPageToFit={true}
                      scrollEnabled={false}
                      source={
                        Object {
                          "html": "<html><head><style>html, body { margin:0; padding:0; overflow:hidden; background-color: transparent; } svg { position:fixed; top:0; left:0; height:100%; width:100% }</style></head><body>n</body></html>",
                        }
                      }
                      style={
                        Array [
                          Object {
                            "flex": 1,
                          },
                          Object {
                            "backgroundColor": "#ffffff",
                          },
                          Array [
                            Object {
                              "backgroundColor": "transparent",
                              "height": 100,
                              "width": 200,
                            },
                            Object {
                              "height": 30,
                              "width": 30,
                            },
                          ],
                        ]
                      }
                    />
                  </View>
                </View>
              </View>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "black",
                    "fontWeight": "700",
                    "maxWidth": 248,
                    "textAlign": "center",
                  }
                }
              >
                categoryName
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
